### ISSUE TITLE
Immutable Apply?

### ISSUE THREAD

**Issue by curran** [ottypes/json0#26]
It would be really great if the library could apply ops such that instead of mutating the original data, a new object is created (with shared structure with the previous data object).

Related:

 * [Immutable Update Patterns](https://redux.js.org/recipes/structuring-reducers/immutable-update-patterns)
 * [Immer](https://github.com/mweststrate/immer) <-- I might try using this with json0

**Comment by curran** [comment]
I discovered an actual implementation of this idea here https://github.com/WeTransfer/json0/commit/00d10c9a0bcfd5ae16314006fa0cfc5c6cf29351

**Comment by curran** [comment]
It turns out `json1` implements immutable apply https://github.com/ottypes/json1/blob/master/test/immutable.coffee#L7

**Comment by josephg** [comment]
Yeah I did it that way there because I totally agree with you - apply should be immutable.

I'd support a PR changing apply to work this way. I don't think we need to pull in an immutable library to do it. Just making a shallow copy on the way down to the modified object should be enough. I'm not sure if inserted data should be cloned in or just moved in. I think if apply is immutable, taking a reference should be fine. (It should be one or the other.)

In the tests the best way to verify this is to recursively call Object.freeze() on all the input data before calling apply. That way if anything is accidentally mutated, the tests will fail.

**Comment by curran** [comment]
FWIW I managed to solve this by monkey patching `apply` and using [immer](https://github.com/immerjs/immer), like this:

```js
import produce from 'immer';
const originalApply = json0.type.apply;
json0.type.apply = (snapshot, op) =>
  produce(snapshot, draftSnapshot => {
    originalApply(draftSnapshot, op);
  });
```
